### ISSUE TITLE
Bindings migration to use AbstractStorageBasedTypeProvider

### ISSUE THREAD

**Issue by lolodomo** [openhab/openhab-addons#14954]
All bindings implementing dynamic thing types, channel types or channel group types are now initialized with a delay of 2 minutes. The thing status is NOT_YET_READY during this delay.
See issue openhab/openhab-core#3394
PR openhab/openhab-core#3407 introduces `AbstractStorageBasedTypeProvider` to reduce this problem. It has now to be used in the 20 bindings with dynamic type providers.

Bindings implementing ThingTypeProvider:

- [ ] digitalstrom
- [ ] homematic
- [ ] mqtt/generic
- [ ] netatmo
- [ ] systeminfo => PR #14501 

Bindings implementing ChannelTypeProvider:

- [ ] bluetooth/generic
- [ ] digitalstrom
- [ ] guntamatic
- [ ] harmonyhub => PR #14507 
- [ ] homematic
- [ ] km200
- [ ] miio
- [ ] mqtt/generic
- [ ] netatmo
- [ ] networkupstools
- [ ] remoteopenhab
- [ ] resol
- [ ] sensibo
- [ ] smartmeter
- [ ] solarwatt
- [ ] tacmi
- [ ] tr064
- [ ] wundergroundupdatereceiver
- [ ] yamahareceiver

Bindings implementing ChannelGroupTypeProvider:

- [ ] harmonyhub => PR #14507 
- [ ] homematic
- [ ] km200
- [ ] mqtt/generic
- [ ] sensibo

**Comment by J-N-K** [comment]
Please note that not all of these suffer from the issue with delayed initialization. It depends on the implementation of the provider. 

Tr064 is not affected, it has a dynamic provider, but it can create all necessary channels when the binding starts, without a storage.

**Comment by J-N-K** [comment]
Please note that not all of these suffer from the issue with delayed initialization. It depends on the implementation of the provider. 

Tr064 is not affected, it has a dynamic provider, but it can create all necessary channels when the binding starts, without a storage.

**Comment by lolodomo** [comment]
> Tr064 is not affected, it has a dynamic provider, but it can create all necessary channels when the binding starts, without a storage.

Interesting. I imagine these channels are not dependent on the things themself.
The number of impacted bindings could be finally largely reduced.
If we have nothing to do for certain of these bindings, that is perfect.

For the remote openHAB binding (I am working on it), I am sure we need the storage.

**Comment by lolodomo** [comment]
@J-N-K : it seems there is a problem. When I call putChannelType, I got many errors of this kind:
```
00:04:16.548 [ERROR] [ore.storage.json.internal.JsonStorage] - Couldn't deserialize value 'org.openhab.core.storage.json.internal.StorageEntry@35e456f3'. Root cause is: Interfaces can't be instantiated! Register an InstanceCreator or a TypeAdapter for this type. Interface name: org.openhab.core.types.StateDescriptionFragment
```

**Comment by J-N-K** [comment]
Yes. We should probably not store `StateDescriptionFragment` but a new `StateDescriptionFragmentEntity` instead.

**Comment by clinique** [comment]
> Please note that not all of these suffer from the issue with delayed initialization. It depends on the implementation of the provider.
> 
> Tr064 is not affected, it has a dynamic provider, but it can create all necessary channels when the binding starts, without a storage.

I have not seen these `WARN` message with Netatmo binding neither.

**Comment by lolodomo** [comment]
> I have not seen these `WARN` messages with Netatmo binding neither.

Me too.
I suppose that this will be OK for dynamic types that are not really dynamic, I mean types that are hardcoded in the binding code but handled as dynamic by the binding ?
I will check the netatmo binding code to understand.

**Comment by lolodomo** [comment]
I am also using the networkupstools binding and did not notice delay at initialization, and warnings so maybe another one that is not impacted.

**Comment by lolodomo** [comment]
@J-N-K / I can confirm that your PR openhab/openhab-core#3598 properly resolved my problem. My PR for the remote openHAB binding is now working well.

**Comment by lolodomo** [comment]
> > I have not seen these `WARN` messages with Netatmo binding neither.
> 
> Me too. I suppose that this will be OK for dynamic types that are not really dynamic, I mean types that are hardcoded in the binding code but handled as dynamic by the binding ? I will check the netatmo binding code to understand.

Ok, the netatmo binding gets its dynamic thing types and channel types from Java enum classes, that is why the thing types and channel types are present at binding startup. No need for the storage in this case.

**Comment by moodyblue** [comment]
I have the 2 minute delay in the following bindings:
 - daikin
 - SmartHome J/HTTP
 - miio (mentioned above)

**Comment by moodyblue** [comment]
> I am also using the networkupstools binding and did not notice delay at initialization

Me too, no delay in that binding

**Comment by jlaur** [comment]
It seems there might be an issue with the Kodi binding as well? I didn't see this with my file-based configuration though, but it was reported here:
https://community.openhab.org/t/openhab-4-0-milestone-discussion/145133/595

```
2023-07-20 18:14:21.100 [WARN ] [core.thing.internal.ThingManagerImpl] - Channel types or config descriptions for thing 'kodi:kodi:9849245e' are missing in the respective registry for more than 120s. In case it does not happen immediately after an upgrade, it should be fixed in the binding.
2023-07-20 18:14:21.103 [WARN ] [core.thing.internal.ThingManagerImpl] - Failed to normalize configuration for thing 'kodi:kodi:9849245e': {thing/channel=Type description kodi:volume for kodi:kodi:9849245e:volume not found, although we checked the presence before.}
```

**Comment by lsiepel** [comment]
Not in the list, but could denonmarantz binding also be affected? See https://github.com/openhab/openhab-addons/issues/15444#issuecomment-1683958966

**Comment by lsiepel** [comment]
@lolodomo I update smart meter. Could you have a look at the PR?

**Comment by lolodomo** [comment]
> [@lolodomo](https://github.com/lolodomo) I update smart meter. Could you have a look at the PR?

I will have to remind me what has to be done because we worked on that 2 years ago.

That being said, we should also check that it was implemented for the matter binding and the zwavejs binding in particular?

**Comment by lsiepel** [comment]
I’ll need to read it all to understand it. Will do. Those two and maybe more might be affected. If all bindings are done, can we think of something to prevent new bindings from wrong implementation? 
Changing the interface or something like that?

Edit: zwavejs creates all types at binding init from initial server response and therefore does not need any storage

**Comment by maniac103** [comment]
I'm currently trying to convert the homematic binding (see [this branch](https://github.com/maniac103/openhab2-addons/tree/hm-type-provider)), but am hitting a road block there: while `AbstractStorageBasedTypeProvider` handles thing types, channel types and channel groups, it doesn't handle config descriptions. Therefore I'm getting nearer, but not yet there:

```
15:01:23.280 [DEBUG] [.core.thing.internal.ThingManagerImpl] - Check result is 'not ready': ThingPrerequisites{thingUID=homematic:HG-HB-UNI-Sen-Wa-Ps:D10DF782-F9E4-390C-CFFC-3DCD6F37A5A9:DBWS000001, thingTypeUID=homematic:HG-HB-UNI-Sen-Wa-Ps, channelTypeUIDs=[homematic:HB-UNI-Sen-Wa-Ps_0_DEVICE_IN_BOOTLOADER, homematic:HB-UNI-Sen-Wa-Ps_0_CONFIG_PENDING, homematic:HB-UNI-Sen-Wa-Ps_1_FILLING_LITER, homematic:HB-UNI-Sen-Wa-Ps_0_DELETE_DEVICE, homematic:HB-UNI-Sen-Wa-Ps_0_RSSI_DEVICE, homematic:HB-UNI-Sen-Wa-Ps_0_RSSI_PEER, homematic:HB-UNI-Sen-Wa-Ps_0_RSSI, homematic:HB-UNI-Sen-Wa-Ps_0_DELETE_DEVICE_MODE, homematic:HB-UNI-Sen-Wa-Ps_0_UPDATE_PENDING, homematic:HB-UNI-Sen-Wa-Ps_0_STICKY_UNREACH, homematic:HB-UNI-Sen-Wa-Ps_0_FIRMWARE, homematic:HB-UNI-Sen-Wa-Ps_0_UNREACH, homematic:HB-UNI-Sen-Wa-Ps_1_FILLING_LEVEL], configDescriptionUris=[]}
```

@J-N-K Was this omitted on purpose, or was it an oversight? I wonder what the correct way to deal with it is. I mainly see two options:
- Extend `AbstractStorageBasedTypeProvider` to cover config descriptions
- Implement the config description storage part in my `HomematicTypeProvider`

FWIW, the config descriptions are built dynamically because they depend on server (CCU) responses, so I can't build them on binding startup or something. I imagine e.g. the Matter and the Zwave-JS bindings having the same kind of problem.

**Comment by marcelGoerentz** [comment]
@lolodomo you can tick the evcc binding

**Comment by lsiepel** [comment]
@maniac103 did not look at the code, but i thought the homematic binding was already adapted or am i mistaken?

**Comment by maniac103** [comment]
@lsiepel Yes, in #19975 (plus some fixups on top)

**Comment by ML19821** [comment]
freeathome: same problem, migration in #21559 (measurements in #21558). The config description gap discussed above does not bite there - the binding's only URI pointed at nothing and is removed in the same PR.
